storage/isolation_manager.rs

```diff
@@ -40,130 +43,256 @@ pub(crate) struct IsolationManager {
+        first_window_relative_index: i64,
+    ) -> Result<(), IsolationError> {
+        let mut window_index = 0;
+        let mut slot_index = (0..TIMELINE_WINDOW_SIZE).find(|si| {
+            match windows[0].get_sequence_number(*si) {
+                Some(seq) => seq.number() >= commit_record.open_sequence_number.number(),
+                None => false
+            }
+        }).unwrap();
+        let start_record = first_window_relative_index + (window_index * TIMELINE_WINDOW_SIZE + slot_index) as i64;
+        for _at in start_record..commit_relative_index {
+            debug_assert!(window_index < windows.len());
+            resolve_concurrent(commit_record, slot_index, &windows[window_index])?;
+            slot_index += 1;
+            if slot_index >= TIMELINE_WINDOW_SIZE {
+                window_index += 1;
+                slot_index = 0;
+            }
+        }
+        debug_assert_eq!(
+            first_window_relative_index + (window_index * TIMELINE_WINDOW_SIZE + slot_index) as i64,
+            commit_relative_index
+        );
+        Ok(())
     }
 
     pub(crate) fn watermark(&self) -> SequenceNumber {
-        self.timeline.watermark()
+        let (window, slot_index) = self.timeline.try_get_window(self.timeline.watermark()).unwrap();
+        match window.get_status(slot_index) {
+            CommitStatus::Applied(seq, _) | CommitStatus::Closed(seq) => seq.clone(),
+            CommitStatus::Validated(_, _) | CommitStatus::Pending(_, _) | CommitStatus::Empty =>
+                unreachable!("The isolation manager must always have the watermark commit record in the window, and it must be either committed or aborted."),
+        }
     }
 
-    pub(crate) fn apply_to_commit_record<F, T>(&self, sequence_number: SequenceNumber, function: F) -> T
+    pub(crate) fn apply_to_commit_record<F, T>(&self, relative_index: i64, function: F) -> T
     where
         F: FnOnce(&CommitRecord) -> T,
     {
-        let shared_window = self.timeline.get_window(sequence_number);
-        let record = shared_window.get_record(sequence_number);
-        function(record)
+        let (window, index) = self.timeline.try_get_window(relative_index).unwrap();
+        let record = match window.get_status(index) {
+            CommitStatus::Validated(_, commit_record) | CommitStatus::Applied(_, commit_record) => commit_record,
+            _ => panic!("apply_to_commit_record called on uncommitted record"), // TODO: Do we want to be able to apply on pending?
+        };
+        // debug_assert_eq!(read_sequence_number, sequence_number);
+        function(&record)
+    }
+
+    pub(crate) fn iterate_commit_status_from_disk<'a, D>(
+        durability_service: &'a D,
+        start_sequence_number: SequenceNumber,
+    ) -> durability::Result<impl Iterator<Item = durability::Result<CommitStatus<'a>>>>
+    where
+        D: DurabilityService,
+    {
+        let mut statuses: HashMap<u128, bool> = HashMap::new();
```

Comment:
Can't we just make `SequenceNumber` have `Hash`? this can map from sequence number directly